newIDE/app/src/Utils/GDevelopServices/Badge.js

```diff
@@ -0,0 +1,107 @@
+// @flow
+import axios from 'axios';
+import { GDevelopUserApi } from './ApiConfigs';
+
+import { type AuthenticatedUser } from '../../Profile/AuthenticatedUserContext';
+
+export const ACHIEVEMENT_FEATURE_FLAG = false;
+
+export const TRIVIAL_FIRST_EVENT = 'trivial_first-event';
+export const TRIVIAL_FIRST_BEHAVIOR = 'trivial_first-behavior';
+export const TRIVIAL_FIRST_PREVIEW = 'trivial_first-preview';
+export const TRIVIAL_FIRST_WEB_EXPORT = 'trivial_first-web-export';
+export const TRIVIAL_FIRST_EXTENSION = 'trivial_first-extension';
+export const TRIVIAL_FIRST_EFFECT = 'trivial_first-effect';
+export const TRIVIAL_FIRST_DEBUG = 'trivial_first-debug';
+
+export type Badge = {|
+  seen: boolean,
+  unlockedAt: string,
+  userId: string,
+  achievementId: string,
+|};
+
+export type Achievement = {|
+  id: string,
+  category: string,
+  name: string,
+  description: string,
+|};
+
+const isAchievementAlreadyClaimed = (
+  badges: Badge[],
+  achievementId: string
+): boolean => {
+  return badges.map(badge => badge.achievementId).includes(achievementId);
+};
+
+const createOrEnsureBadgeForUser = async (
+  authenticatedUser: AuthenticatedUser,
+  achievementId: string
+): Promise<?Badge> => {
+  const {
+    badges,
+    firebaseUser,
+    getAuthorizationHeader,
+    onBadgesChanged,
+  } = authenticatedUser;
+  if (!badges || !firebaseUser) return null;
+  if (isAchievementAlreadyClaimed(badges, achievementId)) {
+    return null;
+  }
+  console.log(`posting achievement ${achievementId}`);
+
+  const userId = firebaseUser.uid;
+  try {
+    const authorizationHeader = await getAuthorizationHeader();
+    const response = await axios.post(
+      `${GDevelopUserApi.baseUrl}/user/${userId}/badge`,
+      {
+        achievementId,
+      },
+      {
+        params: {
+          userId,
+        },
+        headers: {
+          Authorization: authorizationHeader,
+        },
+      }
+    );
+    onBadgesChanged();
+    return response.data;
+  } catch (err) {
+    if (err.response && err.response.status === 409) {
+      console.warn('Badge already exists');
+    } else {
+      throw err;
+    }
+  }
+};
+
+/**
+ * Check if user has already claimed the achievement, to avoid executing
+ * any extra code if that's the case.
+ */
+export const addCreateBadgePreHookIfNotClaimed = (
+  authenticatedUser: AuthenticatedUser,
+  achievementId: string,
+  callback: Function
```

Comment:
I'll try to find if we can type this exactly with Flow.
The risk of using Function is that we lose all type information after calling `addCreateBadgePreHookIfNotClaimed`.

Reply:
@AlexandreSi Give a try at making this generic?

`addCreateBadgePreHookIfNotClaimed<T> = (..., callback: T): T => { ... }`

Reply:
I've searched a bit but it seems complicated.
We always return a function with the same signature as callback. 
Writing this, I realize that we don't return anything, i'll fix this (well typing is useful!)

Reply:
![image](https://user-images.githubusercontent.com/32449369/143047663-41d36e10-8186-4a2c-9597-1917ea94ce12.png)

And found this post https://stackoverflow.com/questions/69005241/cant-return-a-function-from-function-that-returns-a-constrained-generic-type

Reply:
I think flow cannot make sure that the function we return is actually T?

Reply:
I'm giving a try at if we can do this without `// $FlowFixMe` (equivalent to // @ts-ignore), but if not we can:
- still use T in the definition of `addCreateBadgePreHookIfNotClaimed`
- Use // $FlowFixMe inside `addCreateBadgePreHookIfNotClaimed`, so that the only risk, not properly typed stuff is inside the function, but the callers are properly typed including the return type (`Function` is equivalent to `any`, so this spills the beans and we'll lose confidence in all the code calling this).